packages/@aws-cdk/aws-appmesh/lib/virtual-node-listener.ts

```diff
@@ -0,0 +1,366 @@
+ */
+export interface HttpNodeListenerOptions extends VirtualNodeListenerOptions {
+  /**
+   * Timeout for HTTP protocol
+   *
+   * @default - None
+   */
+  readonly timeout?: HttpTimeout;
+}
+
+/**
+ * Represent the GRPC Node Listener prorperty
+ */
+export interface GrpcNodeListenerOptions extends VirtualNodeListenerOptions {
+  /**
+   * Timeout for GRPC protocol
+   *
+   * @default - None
+   */
+  readonly timeout?: GrpcTimeout;
+}
+
+/**
+ * Represent the TCP Node Listener prorperty
+ */
+export interface TcpNodeListenerOptions extends VirtualNodeListenerOptions {
+  /**
+   * Timeout for TCP protocol
+   *
+   * @default - None
+   */
+  readonly timeout?: TcpTimeout;
+}
+
+/**
+ * Represents timeouts for HTTP protocols.
+ */
+export interface HttpTimeout {
+  /**
+   * Represents an idle timeout. The amount of time that a connection may be idle.
+   *
+   * @default - none
+   */
+  readonly idle?: cdk.Duration;
+
+  /**
+   * Represents per request timeout.
+   *
+   * @default - 15 s
+   */
+  readonly perRequest?: cdk.Duration;
+}
+
+/**
+ * Represents timeouts for GRPC protocols.
+ */
+export interface GrpcTimeout {
+  /**
+   * Represents an idle timeout. The amount of time that a connection may be idle.
+   *
+   * @default - none
+   */
+  readonly idle?: cdk.Duration;
+
+  /**
+   * Represents per request timeout.
+   *
+   * @default - 15 s
+   */
+  readonly perRequest?: cdk.Duration;
+}
+
+/**
+ * Represents timeouts for TCP protocols.
+ */
+export interface TcpTimeout {
+  /**
+   * Represents an idle timeout. The amount of time that a connection may be idle.
+   *
+   * @default - none
+   */
+  readonly idle?: cdk.Duration;
+}
+
+/**
+ *  Defines listener for a VirtualNode
+ */
+export abstract class VirtualNodeListener {
+  /**
+   * Returns an HTTP Listener for a VirtualNode
+   */
+  public static http(props: HttpNodeListenerOptions = {}): VirtualNodeListener {
+    return new HttpNodeListener(props);
+  }
+
+  /**
+   * Returns an HTTP2 Listener for a VirtualNode
+   */
+  public static http2(props: HttpNodeListenerOptions = {}): VirtualNodeListener {
+    return new Http2NodeListener(props);
+  }
+
+  /**
+   * Returns an GRPC Listener for a VirtualNode
+   */
+  public static grpc(props: GrpcNodeListenerOptions = {}): VirtualNodeListener {
+    return new GrpcNodeListener(props);
+  }
+
+  /**
+   * Returns an TCP Listener for a VirtualNode
+   */
+  public static tcp(props: TcpNodeListenerOptions = {}): VirtualNodeListener {
+    return new TcpNodeListener(props);
+  }
+
+  /**
+   * Protocol the listener implements
+   */
+  protected abstract protocol: Protocol;
+
+  /**
+   * Port to listen for connections on
+   */
+  protected abstract port: number;
+
+  /**
+   * Health checking strategy upstream nodes should use when communicating with the listener
+   */
+  protected abstract healthCheck?: HealthCheck;
+
+  /**
+   * Represents Listener Timeout
+   */
+  protected abstract timeout?: HttpTimeout;
```

Comment:
None of these are needed. We should strive to minimize the surface area of public classes.